zcash_client_sqlite/src/testing/pool.rs

```diff
@@ -301,167 +302,456 @@ pub(crate) fn send_single_step_proposed_transfer<T: ShieldedPoolTester>() {
+            })
+            .collect();
+
+        assert!(expected_step0_change < expected_ephemeral);
+        assert_eq!(confirmed_sent.len(), 2);
+        assert_eq!(confirmed_sent[0].len(), 2);
+        assert_eq!(
+            confirmed_sent[0][0].0,
+            u64::try_from(expected_step0_change).unwrap()
+        );
+        let (ephemeral_v, to_addr, ephemeral_addr, index) = confirmed_sent[0][1].clone();
+        assert_eq!(ephemeral_v, u64::try_from(expected_ephemeral).unwrap());
+        assert!(to_addr.is_some());
+        assert_eq!(ephemeral_addr, to_addr);
+        assert_eq!(index, Some(expected_index));
+
+        assert_eq!(confirmed_sent[1].len(), 1);
+        assert_matches!(
+            confirmed_sent[1][0].clone(),
+            (sent_v, sent_to_addr, None, None)
+            if sent_v == u64::try_from(transfer_amount).unwrap() && sent_to_addr == Some(tex_addr.encode(&st.wallet().params)));
+
+        (ephemeral_addr.unwrap(), txids)
+    };
 
-    let fee_rule = StandardFeeRule::Zip317;
-    let input_selector = GreedyInputSelector::new(
-        standard::SingleOutputChangeStrategy::new(fee_rule, None, T::SHIELDED_PROTOCOL),
-        DustOutputPolicy::default(),
+    // Each transfer should use a different ephemeral address.
+    let (ephemeral0, txids0) = run_test(&mut st, 0);
+    let (ephemeral1, txids1) = run_test(&mut st, 1);
+    assert_ne!(ephemeral0, ephemeral1);
+
+    let height = add_funds(&mut st, value);
+
+    let ephemeral_taddr = Address::decode(&st.wallet().params, &ephemeral0).expect("valid address");
+    assert_matches!(
+        ephemeral_taddr,
+        Address::Transparent(TransparentAddress::PublicKeyHash(_))
     );
-    let proposal0 = st
-        .propose_transfer(
-            account.account_id(),
-            &input_selector,
-            request0,
+
+    // Attempting to pay to an ephemeral address should cause an error.
+    let proposal = st
+        .propose_standard_transfer::<Infallible>(
+            account_id,
+            StandardFeeRule::Zip317,
             NonZeroU32::new(1).unwrap(),
+            &ephemeral_taddr,
+            transfer_amount,
+            None,
+            None,
+            T::SHIELDED_PROTOCOL,
         )
         .unwrap();
 
-    let min_target_height = proposal0.min_target_height();
-    let step0 = &proposal0.steps().head;
+    let create_proposed_result = st.create_proposed_transactions::<Infallible, _>(
+        account.usk(),
+        OvkPolicy::Sender,
+        &proposal,
+    );
+    assert_matches!(
+        &create_proposed_result,
+        Err(Error::PaysEphemeralTransparentAddress(address_str)) if address_str == &ephemeral0);
+
+    // Simulate another wallet sending to an ephemeral address with an index
+    // within the current gap limit. The `PaysEphemeralTransparentAddress` error
+    // prevents us from doing so straightforwardly, so we'll do it by building
+    // a transaction and calling `store_decrypted_tx` with it.
```

Comment:
I apologize for this test being kind-of long, but the latter parts rely on having done the two proposals above, so it was much easier to extend the test this way.